homeassistant/helpers/script.py

```diff
@@ -309,173 +222,288 @@ def async_log_exception(self, logger, message_base, exception):
+    async def _async_event_step(self):
         """Fire an event."""
-        self.last_action = action.get(CONF_ALIAS, action[CONF_EVENT])
-        self._log("Executing step %s" % self.last_action)
-        event_data = dict(action.get(CONF_EVENT_DATA, {}))
-        if CONF_EVENT_DATA_TEMPLATE in action:
+        self.last_action = self._action.get(CONF_ALIAS, self._action[CONF_EVENT])
+        self._log("Executing step %s", self.last_action)
+        event_data = dict(self._action.get(CONF_EVENT_DATA, {}))
+        if CONF_EVENT_DATA_TEMPLATE in self._action:
             try:
                 event_data.update(
-                    template.render_complex(action[CONF_EVENT_DATA_TEMPLATE], variables)
+                    template.render_complex(
+                        self._action[CONF_EVENT_DATA_TEMPLATE], self._variables
+                    )
                 )
             except exceptions.TemplateError as ex:
-                _LOGGER.error("Error rendering event data template: %s", ex)
+                self._log(
+                    "Error rendering event data template: %s", ex, level=logging.ERROR
+                )
 
-        self.hass.bus.async_fire(action[CONF_EVENT], event_data, context=context)
+        self._hass.bus.async_fire(
+            self._action[CONF_EVENT], event_data, context=self._context
+        )
 
-    async def _async_check_condition(self, action, variables, context):
+    async def _async_condition_step(self):
         """Test if condition is matching."""
-        config_cache_key = frozenset((k, str(v)) for k, v in action.items())
+        config_cache_key = frozenset((k, str(v)) for k, v in self._action.items())
         config = self._config_cache.get(config_cache_key)
         if not config:
-            config = await condition.async_from_config(self.hass, action, False)
+            config = await condition.async_from_config(self._hass, self._action, False)
             self._config_cache[config_cache_key] = config
 
-        self.last_action = action.get(CONF_ALIAS, action[CONF_CONDITION])
-        check = config(self.hass, variables)
-        self._log(f"Test condition {self.last_action}: {check}")
-
+        self.last_action = self._action.get(CONF_ALIAS, self._action[CONF_CONDITION])
+        check = config(self._hass, self._variables)
+        self._log("Test condition %s: %s", self.last_action, check)
         if not check:
             raise _StopScript
 
-    def _async_set_timeout(self, action, variables, context, continue_on_timeout):
-        """Schedule a timeout to abort or continue script."""
-        timeout = action[CONF_TIMEOUT]
-        unsub = None
+    def _log(self, msg, *args, level=logging.INFO):
+        self._script._log(msg, *args, level=level)  # pylint: disable=protected-access
+
+    def _raise(self, msg, *args, exception=None):
+        # pylint: disable=protected-access
+        self._script._raise(msg, *args, exception=exception)
+
+
+class _ScriptRun(_ScriptRunBase):
```

Comment:
Let's name each class to the type it implements. 
```suggestion
class _AsynchronousScriptRun(_ScriptRunBase):
```

Reply:
There's only "legacy" (_LegacyScriptRun) and non-legacy (_ScriptRun). Background & blocking run modes both use _ScriptRun, because it's a function of the caller, not the callee.

And, ultimately, I think that the legacy stuff should get ripped out (after a reasonable deprecation warning period.) At that time _ScriptRunBase can be merged into _ScriptRun to simplify the implementation. I only factored out the common stuff into a base class to make adding legacy support easier and cleaner.